Add Safe and Multicall tools (#1537)
iotaledger/iota-wiki: docs/build/getting-started/multicall.md

---

> description: 'It allows you to bundle multiple calls to various contracts, for both reading and writing data, into a single transaction.'

Reviewer: ```suggestion
description: 'Multicall3 allows you to bundle multiple calls to various contracts, for both reading and writing data, into a single transaction.'
```

Reviewer: The first sentence should not start with it.

Reviewer: The first sentence should not start with it.

Author: @vivekjain23 any reason why this is not yet fixed? I will commit it 👍

image: /img/logo/WASP_logo_dark.png
---

# Multicall3 
## Powering Efficient Batch Transactions on L2 Networks

Multicall3 is a smart contract on the Ethereum Virtual Machine (EVM) that streamlines transactions. It allows you to bundle multiple calls to various contracts, for both reading and writing data, into a single transaction.

> This can be particularly useful when working with decentralized applications that require multiple contract interactions or when batching multiple calls to save on gas costs. To use the Multicall3 contract on Shimmer EVM, the Multicall3 contract is deployed at this address: [0xcA11bde05977b3631167028862bE2a173976CA11](https://explorer.evm.iota.org/address/0xcA11bde05977b3631167028862bE2a173976CA11?tab=write_contract).

Reviewer: ```suggestion
This can be particularly useful when working with decentralized applications that require multiple contract interactions or when batching multiple calls to save on gas costs. You can use the Multicall3 contract on Shimmer EVM deployed at the following address: [0xcA11bde05977b3631167028862bE2a173976CA11](https://explorer.evm.iota.org/address/0xcA11bde05977b3631167028862bE2a173976CA11?tab=write_contract).
```

> This contract offers significant advantages for optimizing transaction processing on L2 networks, also known as Layer 2 scaling solutions. Here's how Multicall3 enhances efficiency:

Reviewer: ```suggestion

## Benefits of Multicall3

This contract offers significant advantages for optimizing transaction processing on L2 networks, also known as Layer 2 scaling solutions. Here's how Multicall3 enhances efficiency:
```

> - **Reduced Gas Costs:** By grouping calls together, Multicall3 minimizes the total gas fees associated with interacting with the blockchain. Gas refers to the computational effort required to execute a transaction on the network.

Reviewer: ```suggestion
### Reduced Gas Costs

Multicall3 minimizes the total gas fees associated with interacting with the blockchain by grouping calls together. 
```

> - **Guaranteed Data Consistency:** Multicall3 fetches data from a single block on the blockchain, ensuring that all the information you receive reflects a consistent state. This eliminates discrepancies that might arise from fetching data across separate transactions.

Reviewer: ```suggestion
### Guaranteed Data Consistency

Multicall3 fetches data from a single block on the blockchain, ensuring that all the information you receive reflects a consistent state. This eliminates discrepancies that might arise from fetching data across separate transactions.
```

- **Faster Execution Speeds:** You no longer need to wait for individual transactions to be confirmed one by one. Multicall3 processes everything efficiently in a single transaction, significantly speeding up the overall process.

> - **Simplified Workflows:** Multicall3 removes the complexity of worrying about the order in which transactions are processed, eliminating potential issues caused by race conditions. Race conditions occur when multiple transactions compete for resources, and the outcome can depend on the order they are executed.

Reviewer: ```suggestion
### Simplified Workflows

Multicall3 removes the complexity of worrying about the order in which transactions are processed, eliminating potential issues caused by race conditions. Race conditions occur when multiple transactions compete for resources, and the outcome can depend on the order in which they are executed.
```

> - **All-or-Nothing Execution:** Multicall3 ensures data integrity by functioning on an all-or-nothing basis. This means that either all the calls within the Multicall3 transaction succeed, or none of them do. This prevents partial execution and maintains data consistency.

Reviewer: ```suggestion
### All-or-Nothing Execution

Multicall3 ensures data integrity by functioning on an all-or-nothing basis. This means that either all the calls within the Multicall3 transaction succeed or none of them do. This prevents partial execution and maintains data consistency.
```

:::tip

> TIP: You can find the ABI of the contract [here](https://www.multicall3.com/abi).

Reviewer: ```suggestion
You can find the ABI of the contract in the [Multicall3 website](https://www.multicall3.com/abi).
```

Reviewer: ![image](https://github.com/iotaledger/iota-wiki/assets/85233773/336348f3-5d89-498a-b2a1-ce3b20ce0f11)

### Function: tryBlockAndAggregate

```solidity 
function tryBlockAndAggregate(bool requireSuccess, Call[] calldata calls)
    public
    payable
    returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData);
```
- Description: This function combines features of tryAggregate and retrieves additional block information.

| Parameter Name | Description                                                                 |
|----------------|-----------------------------------------------------------------------------|
| requireSuccess (bool) | A boolean flag indicating whether all calls must succeed (true) or if failures are allowed (false). |
| calls (Call[] calldata) | An array of Call structs. Each Call struct contains the target contract address (address) and the encoded function call data (bytes). |

- Return Values:
  - blockNumber (uint256): The block number where the calls were executed.
  - blockHash (bytes32): The hash of the block where the calls were executed.
  - returnData (Result[] memory): An array of Result structs, as defined in tryAggregate.

### Function: blockAndAggregate

> - ## Description:

Reviewer: ```suggestion
-  Description:
```

This function is a simplified version of tryBlockAndAggregate that always requires all calls to succeed.
```solidity 
function blockAndAggregate(Call[] calldata calls)
    public
    payable
    returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData);
```

|Parameter Name | Description|
|-----------|---------------------------------------------------------------------
| calls (Call[] calldata)    | An array of Call structs, as defined in aggregate.| 

- Return Values:
  - blockNumber (uint256): The block number where the calls were executed.
  - blockHash (bytes32):  The hash of the block where the calls were executed.
  - returnData (Result[] memory): An array of Result structs, as defined in tryAggregate.

### Function: aggregate3

> - ## Description:

Reviewer: ```suggestion

-  Description:

```

Reviewer: or

```suggestion
#### Description
```

This function aggregates calls similar to aggregate but allows specifying whether failures are allowed for each call individually.

```solidity 
function aggregate3(Call3[] calldata calls) public payable returns (Result[] memory returnData);
```

|Parameter Name | Description|
|-----------|---------------------------------------------------------------------
| calls (Call[] calldata)    | An array of Call3 structs. Each Call3 struct includes the target contract address (address), a flag indicating if failure is allowed (bool allowFailure), and the encoded function call data (bytes callData).| 

- Return Values:

>   - returnData (Result[] memory): An array of Result structs, as defined in tryAggregate.

Reviewer: ```suggestion
  - returnData (Result[] memory): An array of Result structs, as defined in `tryAggregate`.
```

Reviewer: Please wrap function names in ``

### Function: aggregate3Value

> - ## Description:

Reviewer: ```suggestion

-  Description:

```

---
description: 'It allows you to bundle multiple calls to various contracts, for both reading and writing data, into a single transaction.'
image: /img/logo/WASP_logo_dark.png
---

# Multicall3 
## Powering Efficient Batch Transactions on L2 Networks

Multicall3 is a smart contract on the Ethereum Virtual Machine (EVM) that streamlines transactions. It allows you to bundle multiple calls to various contracts, for both reading and writing data, into a single transaction.

> This can be particularly useful when working with decentralized applications that require multiple contract interactions or when batching multiple calls to save on gas costs. To use the Multicall3 contract on Shimmer EVM, the Multicall3 contract is deployed at this address: [0xcA11bde05977b3631167028862bE2a173976CA11](https://explorer.evm.iota.org/address/0xcA11bde05977b3631167028862bE2a173976CA11?tab=write_contract).

Author: ```suggestion
This can be particularly useful when working with decentralized applications that require multiple contract interactions or when batching multiple calls to save on gas costs. To use the Multicall3 contract on IOTA EVM, the Multicall3 contract is deployed at this address: [0xcA11bde05977b3631167028862bE2a173976CA11](https://explorer.evm.iota.org/address/0xcA11bde05977b3631167028862bE2a173976CA11?tab=write_contract).
```

### Function: aggregate3

> - #### Description:

Reviewer: ```suggestion
#### Description:
```
or 
```suggestion
* **Description**:
```

Whichever you prefer, but please apply it to the others, as well as the `Return Values`
